### PR TITLE
PACA-814: Add metric to track latency relative to claim extract date

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-814](https://jira.cms.gov/browse/PACA-814)

**User Story or Bug Summary:**

Add a new BFD pipeline metric to measure the time difference between the MAC extract date and the time BFD receives the data, this is useful in creating monitoring / alerting around are data availability SLI.

---

### What Does This PR Do?

Adds a new `Histogram` metric to track the difference in milliseconds between the time a claim is processed by the ETL pipeline and its extract date.  The extract date field is optional so the metric is only updated when the field is present on a claim.  Since the extract date is just a date the latency is calculated using midnight on that date.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

**Review by KyRobbins — COMMENTED** [review]
Due to no timestamp on the extract date, this metric would seem like it might have a lot of variance if the files don't fall at a consistent time every day, is that going to affect the "usefulness" of it?

**Comment by brianburton** [comment]
> Due to no timestamp on the extract date, this metric would seem like it might have a lot of variance if the files don't fall at a consistent time every day, is that going to affect the "usefulness" of it?

Consumers of the metric will just have to be aware of that limitation.

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java [hunk 1]
```diff
@@ -397,16 +400,24 @@ private long maxSequenceInBatch(Collection<RdaChange<TClaim>> claims) {
...
-      metrics.changeAgeMillis.update(age);
+      final long changeAge = Math.max(0L, nowMillis - changeMillis);
+      metrics.changeAgeMillis.update(changeAge);
+
+      final LocalDate extractDate = claim.getSource().getExtractDate();
+      if (extractDate != null) {
+        final LocalDateTime extractTime = extractDate.atStartOfDay();
+        final long extractMillis = extractTime.toInstant(ZoneOffset.UTC).toEpochMilli();
+        final long extractAge = Math.max(0L, nowMillis - extractMillis);
+        metrics.extractAgeMillis.update(extractAge);
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:419]: If my math is correct, and we want to set an alarm for data received after 2pm eastern on the following day, we'd put our threshold at 43 hours (or 154,800,000 milliseconds).  I believe that would account for the shift from eastern to UTC.  Is that correct?

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java [hunk 2]
```diff
@@ -436,6 +447,8 @@ static class Metrics {
     private final Meter transformFailures;
     /** Milliseconds between change timestamp and current time. */
     private final Histogram changeAgeMillis;
+    /** Milliseconds between extract date and current time. */
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:450]:
```suggestion
    /** Milliseconds between extract date and current time, measures the latency between ingestion and when the MAC processes the data . */
```

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java [hunk 3]
```diff
@@ -436,6 +447,8 @@ static class Metrics {
     private final Meter transformFailures;
     /** Milliseconds between change timestamp and current time. */
```
Comment (ericbuckley) [inline apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:448]:
```suggestion
    /** Milliseconds between change timestamp and current time, measures the latency between BFD ingestion and when RDA ingestion. */
```

File: apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:419]: Both the `extractTime` and the `nowMillis` are in UTC so unless I'm mistaken the difference between them should be a true difference in wall clock time.

Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/sink/direct/AbstractClaimRdaSink.java:418]: Just to be safe I have made a small change to ensure that the `nowMillis` and `extractTime` are both computed using the same time zone.  This way even if the clock passed to the sink's constructor is not in UTC the relative difference will still be the same.

**Review by ericbuckley — APPROVED** [review]

**Review by KyRobbins — APPROVED** [review]

**Review by lsmitchell — APPROVED** [review]

**Pull request merged**